sql: correctly handle aggregation functions above windowing level

nvb opened the pull request:
Fixes #14606.

Previously, aggregation functions above the "windowing level" of a
`windowNode` would never be evaluated by a `groupNode`. This is because
they were never pushed down and added to the `groupNode`'s renders so
they were never properly handled there.

We now replace aggregation functions above `windowFuncHolders` with
`IndexedVars` and add the aggregation functions to the `groupNode`'s
renders. This approach using indirection is similar to 5d26b09, where
index variables (column references) above the "windowing level" in the
expression tree also needed to be pushed below so that they were evaluated
correctly.

nvb committed:
sql: correctly handle aggregation functions above windowing level

Fixes #14606.

Previously, aggregation functions above the "windowing level" of a
`windowNode` would never be evaluated by a `groupNode`. This is because
they were never pushed down and added to the `groupNode`'s renders so
they were never properly handled there.

We now replace aggregation functions above `windowFuncHolders` with
`IndexedVars` and add the aggregation functions to the `groupNode`'s
renders. This approach using indirection is similar to 5d26b09, where
index variables (column references) above the "windowing level" in the
expression tree also needed to be pushed below so that they were evaluated
correctly.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14729)
<!-- Reviewable:end -->

knz commented:
So many thanks for investigating this. Gosh, this area is really hard to reason about (for me at least).
LGTM with a nit

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/window.go, line 320 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhM1rBaZ3OFXLent_7a:-KhM1rBaZ3OFXLent_7b:b-58k5md) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L320)):*
> ```Go
> 	// The number of aggregation functions that need to be replaced with IndexedVars
> 	// is unknown, so we collect them here and bind them to an IndexedVarHelper later.
> 	var aggIdxVars []*parser.IndexedVar
> ```

Use an []int here because `BindIfUnbound` below will discard your newly allocated IndexedVar objects right away.

---

*[pkg/sql/window.go, line 981 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhM2gjZqopdemkQTr-G:-KhM2gjZqopdemkQTr-H:bn9bhmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L981)):*
> ```Go
> // windowNodeAggContainer is a IndexedVarContainer providing indirection for
> // aggregate functions. found above the winowing level. See replaceIndexVarsAndAggFuncs.
> type windowNodeAggContainer struct {
> ```

Why not store the ivarHelper here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-:-KhM2SNUKsBx8L5HeXHB:bea5m6i)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Thanks for looking this over!

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/window.go, line 320 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhM1rBaZ3OFXLent_7a:-KhP_UwyEo3xnEOBweaq:bz1r6h0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L320)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Use an []int here because `BindIfUnbound` below will discard your newly allocated IndexedVar objects right away.
</blockquote></details>

I don't really understand this. We return the allocated `IndexedVar` objects from the `replaceExprsAboveWindowing` visitor function and later bind this `*IndexedVar` to the `Container`. What would I be storing in the []int? 

Is there a way to work with an existing `IndexedVarHelper` if we don't know the number of `IndexedVars` we'll need ahead of time?

---

*[pkg/sql/window.go, line 981 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhM2gjZqopdemkQTr-G:-KhPYbflbUl_RWUEL5_z:b-brdhbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L981)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Why not store the ivarHelper here?
</blockquote></details>

Is there a reason to? You know the `IndexedVar` infrastructure better than I do, but it seems like the helper is only used to bind `IndexedVars` to an `IndexedVarContainer`. Once we've done this, do we need to keep it accessible?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-:-KhPan5o7YGVy-aCaqCj:bpj44cx)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm_cancel:" alt=":lgtm_cancel:" align="absmiddle" src="https://reviewable.io/lgtm_cancel.png" height="20" width="61"/> until I understand what's going on. I suspect there may be a simplification possible.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/sql/window.go, line 289 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR481PHOMhqdvkpW2f:-KhR481QfuhQrX9y3A5s:b-i3qxca) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L289)):*
> ```Go
> // below the windowing level. These expression types are:
> // - IndexedVars: window function evaluation requires completion of any wrapped
> //    plan nodes, so if we did nothing here, all existing IndexedVars from sub-plans
> ```

*from source plans

---

*[pkg/sql/window.go, line 292 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4EKWTUoTrb4q9X2N:-KhR4EKWTUoTrb4q9X2O:b-3k5mfh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L292)):*
> ```Go
> //    would be pointing to their values in the last row of the underlying renderNode.
> //    Clearly, we cannot use the renderNode as the IndexedVarContainer for IndexedVars
> //    above the windowing level.
> ```

Please clarify the term "above the windowing level" throughout. I think this is the most important question in this review.
Provide examples.

---

*[pkg/sql/window.go, line 295 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4Oyh0zfO7M6sE6mX:-KhR4Oyh0zfO7M6sE6mY:bw9yrhy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L295)):*
> ```Go
> // - Aggregate Functions: aggregate functions are handled by the groupNode, which
> //    requires the functions to be present in its renders for proper evaluation.
> //    If an aggregate function is found above the windowing level and we do not push
> ```

Explain "push it down"; provide an example.

---

*[pkg/sql/window.go, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4VpPH0pvYHEZIFII:-KhR4VpPH0pvYHEZIFIJ:b-ducie8) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L299)):*
> ```Go
> //
> // To work around both of these cases, we perform four steps:
> // 1. We add renders for each column referenced by any existing IndexedVar or any
> ```

Explain where (and on which node) the renders are being added.

---

*[pkg/sql/window.go, line 304 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4mLdmZ_zGTmWasMQ:-KhR4mLdmZ_zGTmWasMR:b-pe1erp) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L304)):*
> ```Go
> //    uses the windowNode as an IndexedVarContainer (see windowNodeVarContainer and
> //    windowNodeAggContainer).
> // 3. We buffer all of the column values from the newly added renders while computing
> ```

Explain where the values are buffered.

---

*[pkg/sql/window.go, line 308 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4pN0sjiCL4NiFiiE:-KhR4pN0sjiCL4NiFiiF:b-p833g1) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L308)):*
> ```Go
> // 4. When evaluating windowRenders for each row that contain these new IndexedVars,
> //    the windowNode provides the buffered column value for that row through its
> //    IndexedVarContainer interface.
> ```

Some example with detailed explanation of which values get buffered and computed and where would be welcome here.

---

*[pkg/sql/window.go, line 320 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhM1rBaZ3OFXLent_7a:-KhR2lzjUtJ3jTgk-xGc:b-sif70w) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L320)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I don't really understand this. We return the allocated `IndexedVar` objects from the `replaceExprsAboveWindowing` visitor function and later bind this `*IndexedVar` to the `Container`. What would I be storing in the []int? 

Is there a way to work with an existing `IndexedVarHelper` if we don't know the number of `IndexedVars` we'll need ahead of time?
</blockquote></details>

I was confused both ways. (My comment above does not reflect what I had in mind, and what I had in mind was mistaken -- wrong all around)

The reason why I was confused is that your use case here does not fit the usual use case for indexedvars. I am re-reading the code throughout and have several questions open before I am fully on board with this solution.

---

*[pkg/sql/window.go, line 331 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR5ed2FuAzNIOqLFBU:-KhR5ed2FuAzNIOqLFBV:br8lrne) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L331)):*
> ```Go
> 			case *parser.IndexedVar:
> 				if _, ok := n.varContainer.idxMap[t.Idx]; !ok {
> 					// We add a new render to the wrapped renderNode for each new IndexedVar we
> ```

*to the source renderNode

Please extend this comment to explain why an indexedvar at this level may have a different index than the indexedvar in the renderNode, whereby you motivate the use of the `idxMap` (perhaps use an example too). Perhaps add a comment on the definition of `idxMap` as well.

---

*[pkg/sql/window.go, line 348 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR63gQ0BYaqZPoAxXA:-KhR63gQ0BYaqZPoAxXB:b59p7f4) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L348)):*
> ```Go
> 					s.addRenderColumn(t, ResultColumn{Name: t.String(), Typ: t.ResolvedType()})
> 
> 					aggIVar := parser.NewOrdinalReference(len(aggIdxVars))
> ```

How come you don't need the idxMap here? (if you really don't, please add a comment to explain/illustrate why)

---

*[pkg/sql/window.go, line 955 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR6KaLXyacQn7N0vl0:-KhR6KaLXyacQn7N0vl1:bghhsy6) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L955)):*
> ```Go
> 
> // windowNodeVarContainer is a IndexedVarContainer providing indirection for
> // IndexedVars found above the winowing level. See replaceIndexVarsAndAggFuncs.
> ```

Detail the fields with comments.

---

*[pkg/sql/window.go, line 966 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR6NSKy87rZBjIr-Rb:-KhR6NSKy87rZBjIr-Rc:b-vusjhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L966)):*
> ```Go
> 	idx int, ctx *parser.EvalContext,
> ) (parser.Datum, error) {
> 	return vc.n.wrappedIndexedVarVals.At(vc.n.curRowIdx)[vc.idxMap[idx]].Eval(ctx)
> ```

Decompose and add comments to explain what is going on here.

---

*[pkg/sql/window.go, line 981 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhM2gjZqopdemkQTr-G:-KhR26rUQNtmoqA-NbmL:bkgae4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L981)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is there a reason to? You know the `IndexedVar` infrastructure better than I do, but it seems like the helper is only used to bind `IndexedVars` to an `IndexedVarContainer`. Once we've done this, do we need to keep it accessible?
</blockquote></details>

On second read, indeed it is not necessary here.

---

*[pkg/sql/window.go, line 991 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR6TqJrp3MTKpmcSLN:-KhR6TqJrp3MTKpmcSLO:b-vusjhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L991)):*
> ```Go
> 	idx int, ctx *parser.EvalContext,
> ) (parser.Datum, error) {
> 	return ac.n.wrappedIndexedVarVals.At(ac.n.curRowIdx)[idx+ac.rowOffset].Eval(ctx)
> ```

Decompose and add comments to explain what is going on here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-:-KhR6YA94D_DUgLeW7gL:b-qd79mv)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/sql/window.go, line 955 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR6KaLXyacQn7N0vl0:-KhR6j7xe7f16jMAkO0j:b-648rta) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L955)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Detail the fields with comments.
</blockquote></details>

Also, "windowing"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Added a number of clarifications and examples. Let me know if something's still unclear or if you see any possible simplifications.

---

Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/sql/window.go, line 289 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR481PHOMhqdvkpW2f:-KhapuGwUNeAbCvi64Yj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L289)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

*from source plans
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 292 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4EKWTUoTrb4q9X2N:-KhazknQ6eleHwhkYg66:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L292)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please clarify the term "above the windowing level" throughout. I think this is the most important question in this review.
Provide examples.
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 295 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4Oyh0zfO7M6sE6mX:-KhazkIl7PRFxkMVz5tr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L295)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Explain "push it down"; provide an example.
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4VpPH0pvYHEZIFII:-Khazjoi0m7Z9-jIeLIT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L299)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Explain where (and on which node) the renders are being added.
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 304 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4mLdmZ_zGTmWasMQ:-KhaxWHViGMESW90DX6W:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L304)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Explain where the values are buffered.
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 308 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4pN0sjiCL4NiFiiE:-KhaxHeXP4ttD4V6z23V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L308)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Some example with detailed explanation of which values get buffered and computed and where would be welcome here.
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 331 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR5ed2FuAzNIOqLFBU:-KharJFOUI3xVlC1BioW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L331)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

*to the source renderNode

Please extend this comment to explain why an indexedvar at this level may have a different index than the indexedvar in the renderNode, whereby you motivate the use of the `idxMap` (perhaps use an example too). Perhaps add a comment on the definition of `idxMap` as well.
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 348 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR63gQ0BYaqZPoAxXA:-KhaqfTkKPy1oSyPkLZM:b-6zf2ha) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L348)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

How come you don't need the idxMap here? (if you really don't, please add a comment to explain/illustrate why)
</blockquote></details>

idxMap was an optimization. Added a comment to clarify.

---

*[pkg/sql/window.go, line 955 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR6KaLXyacQn7N0vl0:-KharlIxcgKs885pXdp_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L955)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Also, "windowing"
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 966 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR6NSKy87rZBjIr-Rb:-Khas-gH-b78Inwicn48:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L966)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Decompose and add comments to explain what is going on here.
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 991 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR6TqJrp3MTKpmcSLN:-KhasrLvIZ-VgUNIDEdd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L991)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Decompose and add comments to explain what is going on here.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-:-KhazlyIUr3zU08_2GRi:b7ref19)*
<!-- Sent from Reviewable.io -->

knz commented:
Yes, this works much better for me. Thank you for the much improved and more informative comments. Just some nits remaining.

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/window.go, line 295 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4Oyh0zfO7M6sE6mX:-KhbPk8qYR3XK-IDk3mS:b-opfmat) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L295)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Ok now this is clearer but I do not think that the term "push down" naturally conveys what you want it to mean.
What I understand is that the sub-expression containing the aggregate function call must be migrated below the windowing level to the group node, and replaced above the windowing level by an indexedvar that refers to the result of the aggregation stage. 
So first of all you can add an example of the substitution, e.g. say " `SELECT max(x) + row_number() OVER () FROM t` is replaced by `SELECT y + row_number() OVER () FROM (SELECT MAX(x) AS y , ... FROM t)` "

Then actually any existing equivalent aggregation will do for the purpose of converting to an indexed var; this phrasing now reveals that you can avoid adding a new aggregation to the group node if that aggregation is already performed. i.e. `select max(x) + row_number() over (), max(x) from t` should be replaced by `select y + row_number(), y from (select max(x) as y, ... from t)`

---

*[pkg/sql/window.go, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4VpPH0pvYHEZIFII:-KhbQqZzKeroJ1h0Yozj:b1eql17) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L299)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Where are the renders added?

---

*[pkg/sql/window.go, line 304 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4mLdmZ_zGTmWasMQ:-KhbR06jmaPnZ6wP2_jH:b-uxg6d0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L304)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

The sentence is difficult to read. Try this: "The results are computed by the source node for the newly added renders; the window node buffers these results in the wrappedIndexedVarVals RowContainer ..."

---

*[pkg/sql/window.go, line 955 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR6KaLXyacQn7N0vl0:-KhbRsoN4tAtUgJ7YalI:b2e4s3g) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L955)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

"winowing" -> "windowing"

---

*[pkg/sql/window.go, line 289 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhbGirwcn2_olUYnE0P:-KhbGirxtlHIOMazTqn2:b-7qwxae) ([raw file](https://github.com/cockroachdb/cockroach/blob/c622e696415651bd02029762f08bc5ce88bc112f/pkg/sql/window.go#L289)):*
> ```Go
> // expressions will be evaluated after each of the window functions are run, so
> // we refer to them as happening above the "windowing level". In other words, we
> // let the source plan execute to completion below the windowing level, we the
> ```

we *then

---

*[pkg/sql/window.go, line 292 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhbPCXNUOyrmhq0gRFH:-KhbPCXNUOyrmhq0gRFI:bbqyuap) ([raw file](https://github.com/cockroachdb/cockroach/blob/c622e696415651bd02029762f08bc5ce88bc112f/pkg/sql/window.go#L292)):*
> ```Go
> // compute the results of each window function for each row, and finally we continue
> // evaluating the windowRenders with each window function's respective result for
> // that row above the windowing level.
> ```

I like this explanation, however now it raises another question: since the window functions are evaluated at every row, why do we extract them from the surrounding expression at all?
Why not evaluate the surrounding expression normally (i.e. using `Eval`) and have the FuncExpr thing for the window function compute that window function's result for the current row?

Is it because window function have state? If so, is it not possible to maintain the state externally and link it (e.g. using a reference) from the AST node so that its Eval method can do its work? (instead of the complexity of the current two-stage evaluation)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-:-KhbS6w0LMK4cZC0k6QY:b-1hue3k)*
<!-- Sent from Reviewable.io -->

nvb committed:
sql: attempt to merge all added renders in windowNode

This change uses `renderNode`'s `addOrMergeRender` method to attempt to
merge all renders added for window definition partition expressions,
window definition orderings, indexed vars, and aggregate functions. This
addresses a couple of TODOs.

nvanbenschoten commented:
Review status: 0 of 4 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/window.go, line 295 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4Oyh0zfO7M6sE6mX:-Khut_HOR3DMQXRovAqX:bhtpeiw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L295)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Ok now this is clearer but I do not think that the term "push down" naturally conveys what you want it to mean.
What I understand is that the sub-expression containing the aggregate function call must be migrated below the windowing level to the group node, and replaced above the windowing level by an indexedvar that refers to the result of the aggregation stage. 
So first of all you can add an example of the substitution, e.g. say " `SELECT max(x) + row_number() OVER () FROM t` is replaced by `SELECT y + row_number() OVER () FROM (SELECT MAX(x) AS y , ... FROM t)` "

Then actually any existing equivalent aggregation will do for the purpose of converting to an indexed var; this phrasing now reveals that you can avoid adding a new aggregation to the group node if that aggregation is already performed. i.e. `select max(x) + row_number() over (), max(x) from t` should be replaced by `select y + row_number(), y from (select max(x) as y, ... from t)`


</blockquote></details>

Yes, your understanding if correct. I've moved from the term "push down" to "migrate below", which I agree is a little more suggestive.

You're also correct about the possibility of avoiding new aggregation renders for duplicate aggregation functions. I made the change for all cases where windowNode added renders (window definition partition expressions, window definition orderings, indexed vars, and aggregate functions). This addressed a few TODOs and allowed for a bit of cleanup.

This also got me thinking. I don't see any reason why we don't deduplicating aggregation functions in general at the `groupNode` level as well, such that `SELECT stddev(x), 2*stddev(x) FROM t` performs only a single aggregation. This should be pretty easy to do.

---

*[pkg/sql/window.go, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4VpPH0pvYHEZIFII:-KhusMwAkglmqZHU-17F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L299)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Where are the renders added?
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 304 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4mLdmZ_zGTmWasMQ:-KhuseexqMe_cDNWSh22:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L304)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The sentence is difficult to read. Try this: "The results are computed by the source node for the newly added renders; the window node buffers these results in the wrappedIndexedVarVals RowContainer ..."
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 955 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR6KaLXyacQn7N0vl0:-KhushuQDf99MDvUmc2B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L955)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"winowing" -> "windowing"
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 289 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhbGirwcn2_olUYnE0P:-Khus7B2n6Mf-yQGnpS0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c622e696415651bd02029762f08bc5ce88bc112f/pkg/sql/window.go#L289)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

we *then
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 292 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhbPCXNUOyrmhq0gRFH:-KhuwFYyd73Nhm8aS8Ua:b-7mhvtc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c622e696415651bd02029762f08bc5ce88bc112f/pkg/sql/window.go#L292)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I like this explanation, however now it raises another question: since the window functions are evaluated at every row, why do we extract them from the surrounding expression at all?
Why not evaluate the surrounding expression normally (i.e. using `Eval`) and have the FuncExpr thing for the window function compute that window function's result for the current row?

Is it because window function have state? If so, is it not possible to maintain the state externally and link it (e.g. using a reference) from the AST node so that its Eval method can do its work? (instead of the complexity of the current two-stage evaluation)
</blockquote></details>

The result of the window function can be different for each row, but to compute any row we need to have access to all rows at the same time. This is why `windowNode`, like `sortNode`, requires a stream synchronization (there's a proper term for this, but it's escaping me).

The issue with what you're suggesting about the Eval method is that window frames are computed on different partitions of rows which are not always sorted in the same order. State needs to be maintained within these partitions as each row is being evaluated, in order, per window function. The general workflow is 
```
for each window function
    compute partitions
    for each partition
        sort partition
        evaluate window frame in sorted partition per cell, keeping track of peer groups
```
I can't easily see a way of pulling any of the work done in `computeWindows` into each Eval method without some combination of 1. doing much more work, 2. making the code less readable, 3. caching a lot of work and making the external state very unclear. In fact, I think we'd end up just doing exactly what we do now on the first call to `Eval` for a given `windowFuncHolder`, and caching the rest of the results for later, so the end result would be the same as what we have now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo the integration of your explanation in a comment.

---

Reviewed 1 of 1 files at r3, 4 of 4 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/row_container.go, line 213 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhveGquXCUjk86eSY2Q:-KhveGquXCUjk86eSY2R:b-d7udui) ([raw file](https://github.com/cockroachdb/cockroach/blob/269959b9854cf4d54f4fa4bcf36a3421285e9894/pkg/sql/row_container.go#L213)):*
> ```Go
> 
> // Silence unused warning for *RowContainer.NumCols.
> var _ = (*RowContainer).NumCols
> ```

Uh I do not understand this. If the method is unused, why not simply remove it entirely.

---

*[pkg/sql/window.go, line 295 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhR4Oyh0zfO7M6sE6mX:-Khvdvg5UqEtpVuwMTuZ:b3f0a05) ([raw file](https://github.com/cockroachdb/cockroach/blob/3398ba3e17e02f784f4e6d24bd22cf14450e4114/pkg/sql/window.go#L295)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yes, your understanding if correct. I've moved from the term "push down" to "migrate below", which I agree is a little more suggestive.

You're also correct about the possibility of avoiding new aggregation renders for duplicate aggregation functions. I made the change for all cases where windowNode added renders (window definition partition expressions, window definition orderings, indexed vars, and aggregate functions). This addressed a few TODOs and allowed for a bit of cleanup.

This also got me thinking. I don't see any reason why we don't deduplicating aggregation functions in general at the `groupNode` level as well, such that `SELECT stddev(x), 2*stddev(x) FROM t` performs only a single aggregation. This should be pretty easy to do.
</blockquote></details>

Yes I agree! :)

---

*[pkg/sql/window.go, line 292 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhbPCXNUOyrmhq0gRFH:-KhvdoEegf4EMSPwNAr-:b-9igx4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c622e696415651bd02029762f08bc5ce88bc112f/pkg/sql/window.go#L292)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

The result of the window function can be different for each row, but to compute any row we need to have access to all rows at the same time. This is why `windowNode`, like `sortNode`, requires a stream synchronization (there's a proper term for this, but it's escaping me).

The issue with what you're suggesting about the Eval method is that window frames are computed on different partitions of rows which are not always sorted in the same order. State needs to be maintained within these partitions as each row is being evaluated, in order, per window function. The general workflow is 
```
for each window function
    compute partitions
    for each partition
        sort partition
        evaluate window frame in sorted partition per cell, keeping track of peer groups
```
I can't easily see a way of pulling any of the work done in `computeWindows` into each Eval method without some combination of 1. doing much more work, 2. making the code less readable, 3. caching a lot of work and making the external state very unclear. In fact, I think we'd end up just doing exactly what we do now on the first call to `Eval` for a given `windowFuncHolder`, and caching the rest of the results for later, so the end result would be the same as what we have now.
</blockquote></details>

Thank you for explaining. This makes so much more sense.

Also, please integrate your explanation in the comment. :)

---

*[pkg/sql/window.go, line 184 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhveccWIgX7j7IJncAr:-KhveccWIgX7j7IJncAs:bkgqd0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/269959b9854cf4d54f4fa4bcf36a3421285e9894/pkg/sql/window.go#L184)):*
> ```Go
> 				return err
> 			}
> 			windowFn.partitionIdxs = s.addOrMergeRenders(cols, exprs, true)
> ```

:heart_eyes_cat: :100: thank you!!!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-:-Khvf-UuMLjKZIWB-zGb:b5himeg)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/row_container.go, line 213 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhveGquXCUjk86eSY2Q:-KhxeP9xyFGIsl5El5MB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/269959b9854cf4d54f4fa4bcf36a3421285e9894/pkg/sql/row_container.go#L213)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Uh I do not understand this. If the method is unused, why not simply remove it entirely.
</blockquote></details>

Done.

---

*[pkg/sql/window.go, line 292 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729#-KhbPCXNUOyrmhq0gRFH:-KhxjFHoBVqznQpPdY0J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c622e696415651bd02029762f08bc5ce88bc112f/pkg/sql/window.go#L292)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Thank you for explaining. This makes so much more sense.

Also, please integrate your explanation in the comment. :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14729)*
<!-- Sent from Reviewable.io -->
